File: backend/libs/nova_contracts/src/nova_contracts/recorder.py
"""Tick recorder setting (D54): mirrors `frontend/packages/contracts/src/recorder.ts`."""

from typing import Annotated, Literal

from pydantic import Field

from nova_contracts.common import Contract, Id, UtcDateTime
from nova_contracts.universe import Symbol

# Kite streams at most this many instruments over one WebSocket.
MAX_RECORDER_SYMBOLS = 3000

RecorderState = Literal["off", "waiting", "recording", "no_login"]


class RecorderSettingsUpdate(Contract):
    """Body of `PUT /broker/recorder`. `symbols` empty = every stock synced with Kite."""

    enabled: bool
    symbols: Annotated[list[Symbol], Field(max_length=MAX_RECORDER_SYMBOLS)]


class RecorderSettings(RecorderSettingsUpdate):
    """`state`: off; waiting (outside 09:15-15:30 IST on weekdays); recording (`jobId` is the
    running `tick_record` job); no_login (market hours but no live Kite session)."""

    state: RecorderState
    job_id: Id | None
    updated_at: UtcDateTime

File: frontend/packages/contracts/src/recorder.ts
import { z } from "zod";
import { IdSchema, UtcDateTimeSchema } from "./common";
import { UniverseSymbolSchema } from "./universe";

/** Kite streams at most this many instruments over one WebSocket. */
export const MAX_RECORDER_SYMBOLS = 3000;

/**
 * `off`; `waiting` (outside 09:15–15:30 IST on weekdays); `recording` (`jobId` is the running
 * `tick_record` job); `no_login` (market hours but no live Kite session).
 */
export const RecorderStateSchema = z.enum(["off", "waiting", "recording", "no_login"]);
export type RecorderState = z.infer<typeof RecorderStateSchema>;

/** Body of `PUT /broker/recorder` (D54). `symbols` empty = every stock synced with Kite. */
export const RecorderSettingsUpdateSchema = z.strictObject({
  enabled: z.boolean(),
  symbols: z
    .array(UniverseSymbolSchema)
    .max(MAX_RECORDER_SYMBOLS, `At most ${MAX_RECORDER_SYMBOLS} stocks`),
});
export type RecorderSettingsUpdate = z.infer<typeof RecorderSettingsUpdateSchema>;

export const RecorderSettingsSchema = z.strictObject({
  ...RecorderSettingsUpdateSchema.shape,
  state: RecorderStateSchema,
  jobId: IdSchema.nullable(),
  updatedAt: UtcDateTimeSchema,
});
export type RecorderSettings = z.infer<typeof RecorderSettingsSchema>;
